server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateDataStreamService.java

```diff
@@ -277,7 +277,7 @@ static ClusterState createDataStream(
             isSystem,
             template.getDataStreamTemplate().isAllowCustomRouting(),
             indexMode,
-            lifecycle
+            lifecycle == null && DataStreamLifecycle.isFeatureEnabled() ? DataStreamLifecycle.DEFAULT : lifecycle
```

Comment:
> I am wondering if this would be a surprising behaviour though and the user did want to opt out of all management systems by removing the one mechanism that managed their data stream from the one place that mechanism is configured (perhaps another in-house/custom management system is to-be used here)

This is a good argument. I think in the end, it depends on our narrative and where do we want to put emphasis. For example:
- a data stream is always managed by default by data stream lifecycle, unless otherwise specified by explicitly disabling the lifecycle or by using ILM, vs
- when created a data stream that does not have an explicit lifecycle management (ILM or data stream lifecycle) configured will get the default data stream lifecycle. You can opt-out by setting the lifecycle to be  `"lifecycle": {"enabled": false}` or removing the lifecycle all together via the `DELETE ...` API.

Which narrative do we think is friendlier or more intuitive to our users?

Reply:
IMO the second one is what we should strive for (note that we'll not _always_ manage a data stream - we are giving our users options to end up with unmanaged data streams - NB the default management refers only to data stream creation time)

Reply:
I can see the benefit of both, but what about mixed-mode data streams where an ILM policy was used and is no longer, or is now used but not for all indices? Does that change anything for these two options?

Reply:
This remains a tricky one. In this comment, we will summarise our dilemma and the cases in which it applies. @dakrone I am hoping to answer the mixed mode question with this too.

_Disclaimer: Adding a default data stream lifecycle to the new data streams that are not managed by any lifecycle is the requirement of this PR so we do not list it below. The same goes when a user has configured ILM and a data stream lifecycle. We only focus, on how an ILM policy affects the addition of the default data stream lifecycle to a new data stream._

**The options**

1. We always add a default data stream lifecycle.
2. We add a default data stream lifecycle unless **there is an ILM policy**.

We are going to list here the scenarios where the choice above differentiates what the user will experience.

**Scenario 1**
A user has created a data stream that is managed by ILM, **then they remove ILM**:

1. their data stream keeps being managed by elasticsearch, meaning it gets rolled over and tail force merged.
2. their data stream remains unmanaged.

We argue that for this scenario, the first option is better. This enhancement supports the idea that we become more opinionated about data streams having some basic maintenance tasks running on the background, that's exactly what the default lifecycle does. So, having the default data stream lifecycle kick in when a user removes ILM seems to be inline with this idea. Making the user go out of their way to disable the lifecycle, means that it will not be disabled by accident but only when intended, so the user knows what they are doing.

**Scenario 2**
A user has created a data stream that is managed by ILM, **then they flip the switch prefer_ilm to `false`**:

1.  their data stream stops being managed by ILM but they start being managed by the default data stream lifecycle
2. their data stream keeps being managed by ILM, because that's the only lifecycle configuration available.

In this scenario,  we believe that the option 1 is having a not desirable side effect,  ILM gets skipped over a data stream lifecycle that the user did not configure. However, since this requires an explicit change in the flag we believe we can document this correctly and alleviate any surprises.

**Conclusion** 
We believe that the differences between the two options are subtle and only affect a small part of our use cases. The choice depends on how important we think it is to rollover and tail merge a data stream. We propose to move on with always adding the default data stream lifecycle, trying to "protect" a user from accidentally getting a big backing index. The way to disable all lifecycle configuration will be well documented in case it's needed.

If there are any objections we still have time to change this.

Reply:
> We propose to move on with always adding the default data stream lifecycle, trying to "protect" a user from accidentally getting a big backing index.

This sounds good to me, and great job summarizing this in the options above Mary, it's really helpful!